Segment v2 stream load core dump(#2037)

wangbo opened the pull request:
see #2037
1, fix mem leak when calling string builder.get_dictionary_page;
2, fix delete invalid mem addr in bitshuffleBuilder when no array grow happends
     when bitshuffleBuilder didn't grow array, the data page which not use new to allocate will be 
     returned  to ColumnWriter.
     When ColumnWriter destructs, the data page will be deleted,this causes core dump

imay commented on be/src/olap/rowset/segment_v2/rle_page.h:
```diff
@@ -129,7 +124,26 @@ class RlePageBuilder : public PageBuilder {
+
 private:
+
+    Slice _finish(BuilderReturnType builderReturnType) {
```
```suggestion
    Slice _finish(BuilderReturnType type) {
```
In c++, we use lowercase and underline for variable names

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/column_writer.cpp:
```diff
@@ -219,7 +219,9 @@ Status ColumnWriter::write_data() {
         _page_builder->get_dictionary_page(&dict_page);
```
unique_ptr is preferred.
```
std::unique_ptr<char[]> data(dict_page.data)
```

wangbo replied:
👌

imay reviewed (changes requested):
I don't know why the second problem will be fixed by this PR.

wangbo commented:
> I don't know why the second problem will be fixed by this PR.
For page produced by BitShuffleBuilder,ColumnWriter calls finish() to get Slice,that is calling BitShuffleBuilder.fastring->data(),but fastring  may return a  data array which not using new to create when not grow happens.
This pr using BitShuffleBuilder's finish_and_release method to produce page,that is calling BitShuffleBuilder's fastring.release,this page can be deleted by Page's Destructor;

> why create finish_and_release method?
The situation for PageBuilder in ColumnWriter is different from other situations,
Because all data produced by PageBuilder will be kept in ColumnWriter,ColumnWriter maintains the pages and delete at some time,it seems that calling PageBuilder.finish and then calling release() is ok;
But calling PageBuilder.finish means calling PageBuilder.faststring.finish,this comes back to the problem that when faststring not group,ColumnWriter will get a slice'data which not using new to create,so we must calling fastring.release to get data in ColumnWriter'context;
Calling PageBuilder.finish_and_release means calling fastring.release to get data

wangbo commented on be/src/olap/rowset/segment_v2/page_builder.h:
```diff
@@ -79,6 +84,12 @@ class PageBuilder {
+    // Return a Slice which represents the encoded data of current page,and release the slice
+    // caller need to delete the returned slice
+    virtual Slice finish_and_release() {
+        return nullptr;
```
👌

imay commented:
> @imay
> 
> > I don't know why the second problem will be fixed by this PR.
> 
> For page produced by BitShuffleBuilder,ColumnWriter calls finish() to get Slice,that is calling BitShuffleBuilder.fastring->data(),but fastring may return a data array which not using new to create when not grow happens.
> This pr using BitShuffleBuilder's finish_and_release method to produce page,that is calling BitShuffleBuilder's fastring.release,this page can be deleted by Page's Destructor;
> 
> > why create finish_and_release method?
> 
> The situation for PageBuilder in ColumnWriter is different from other situations,
> Because all data produced by PageBuilder will be kept in ColumnWriter,ColumnWriter maintains the pages and delete at some time,it seems that calling PageBuilder.finish and then calling release() is ok;
> But calling PageBuilder.finish means calling PageBuilder.faststring.finish,this comes back to the problem that when faststring not group,ColumnWriter will get a slice'data which not using new to create,so we must calling fastring.release to get data in ColumnWriter'context;
> Calling PageBuilder.finish_and_release means calling fastring.release to get data

@wangbo 
I see and thanks for your reply.
If then I think we have two choices to fix this problem.

1. We make BitShuffleBuilder finish's return slice is same with what it will release. Then there is no need to add finish_and_release.
2. We remove release interface and keep finish and finish_and_release. Because if we can't keep release and finish consistent, it will lead to other problems.

kangpinghuang commented on be/src/olap/rowset/segment_v2/bitshuffle_page.h:
```diff
@@ -139,8 +139,12 @@ class BitshufflePageBuilder : public PageBuilder {
+
 private:
-    Slice _finish(int final_size_of_type) {
+    Slice _finish(int final_size_of_type, BuilderReturnType builderReturnType) {
```
```suggestion
    Slice _finish(int final_size_of_type, BuilderReturnType type) {
```

kangpinghuang commented on be/src/olap/rowset/segment_v2/page_builder.h:
```diff
@@ -79,6 +84,12 @@ class PageBuilder {
 
+    // Return a Slice which represents the encoded data of current page,and release the slice
+    // caller need to delete the returned slice
+    virtual Slice finish_and_release() {
```
I think we can merge release to finish?

wangbo replied:
In some situation, caller just copy the page returned by PageBuilder, so there is no need for caller to release the page,Such as in BinaryDictPageBuilder.finish,BinaryDictPageBuilder's _data_page_builder just return data page is ok,because BinaryDictPageBuilder just copy the page and won't maintain the page

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_plain_page.h:
```diff
@@ -117,12 +109,34 @@ class BinaryPlainPageBuilder : public PageBuilder {
     }
 
 private:
+    Slice _finish(BuilderReturnType returnType) {
```
```suggestion
    Slice _finish(BuilderReturnType return_type) {
```
use _ stype

kangpinghuang commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -77,7 +77,22 @@ class BinaryDictPageBuilder : public PageBuilder {
+     Slice finish_and_release() override;
+
 private:
+    Slice _finish(BuilderReturnType builderReturnType) {
```
```suggestion
    Slice _finish(BuilderReturnType type) {
```

wangbo commented:
> 1. We make BitShuffleBuilder finish's return slice is same with what it will release. Then there is no need to add finish_and_release.
I think it's better to keep all build method having the same semantics;And other PageBuilder using fastring as buffer may meet the same question.Just fix BitShuffleBuilder is not enough.

> 2. We remove release interface and keep finish and finish_and_release. Because if we can't keep release and finish consistent, it will lead to other problems.
There is two situation to get PageBuilder's page:
1. Caller maintains page,  caller must call finish and release,it seems that using release to get result is better.
2. PageBuilder maintains page,caller using page just for copy and read,caller call finish is enough.
And I don't understand what does "Because if we can't keep release and finish consistent, it will lead to other problems" mean

imay commented:
> > 1. We make BitShuffleBuilder finish's return slice is same with what it will release. Then there is no need to add finish_and_release.
> >    I think it's better to keep all build method having the same semantics;And other PageBuilder using fastring as buffer may meet the same question.Just fix BitShuffleBuilder is not enough.
> 
> > 1. We remove release interface and keep finish and finish_and_release. Because if we can't keep release and finish consistent, it will lead to other problems.
> >    There is two situation to get PageBuilder's page:
> 
>     1. Caller maintains page,  caller must call finish and release,it seems that using release to get result is better.
> 
>     2. PageBuilder maintains page,caller using page just for copy and read,caller call finish is enough.
>        And I don't understand what does "Because if we can't keep release and finish consistent, it will lead to other problems" mean

If you want to keep all methods, what's the "release" function means? What is to be released through this function? And for BitShuffleBuilder, is there any problem to use release?

wangbo commented:
@imay 
1. I think move release and add finish_and_release is ok
2. BitShuffleBuilder using fastring,other PageBuilder using faststring too,here I'm not quite sure just other PageBuilder may has the same problem,so it's better to fix them all;
3. I just don't understand what does 'keep release and finish consistent' mean,that means pageBuilder.finish using faststring.release to return?

imay commented:
> @imay
> 
>     1. I think move release and add finish_and_release is ok
> 
>     2. BitShuffleBuilder using fastring,other PageBuilder using faststring too,here I'm not quite sure just other PageBuilder may has the same problem,so it's better to fix them all;

When we determine these interfaces, we should not consider the specific implementation. So first of all, we should discuss to agree on what's the meaning of all function interfaces.

> 
>     3. I just don't understand what does 'keep release and finish consistent' mean,that means pageBuilder.finish using faststring.release to return?

What I want to express is that the `release` content is what `finish` function returns. Class user don't care about its implementation detail. If we can make BitShufflePageBuilder work in this way, then do we really need to add the `finish_and_release` interface?

gaodayue commented:
My suggestion is to combine `Slice finish()` and `Slice release()` into `OwnedSlice finish()`. `OwnedSlice` owns slice's data and is a move-only type. This will make ownership transfer clear and remove all potential bugs.

imay commented:
> My suggestion is to combine `Slice finish()` and `Slice release()` into `OwnedSlice finish()`. `OwnedSlice` owns slice's data and is a move-only type. This will make ownership transfer clear and remove all potential bugs.

This is an option. But if we use adopt it, user can call `finish` only one time. I'm not sure if it is convenient for all use scenarios.

wangbo commented:
>We make BitShuffleBuilder finish's return slice is same with what it will release. Then there is no need to add finish_and_release.
1. This way break PageBuilder.finish 's semantics,when caller calls finish and release, BitShufflePageBuilder will release faststring twice,this would make memory leak;Or caller need to know the BitShufflePageBuilder.finish because it's different from other PageBuilder's finish
2. Other PageBuilder use faststring as buffer too,So they may meet the same problem with BitShuffleBuilder,just fix BitShuffleBuilder may be not enough.

>We remove release interface and keep finish and finish_and_release. Because if we can't keep release and finish consistent, it will lead to other problems.

I prefer this way;when caller just use page returned by PageBuilder to copy,they call finish.
When caller maintains page,they call release_and_finish

>My suggestion is to combine Slice finish() and Slice release() into OwnedSlice finish(). OwnedSlice owns slice's data and is a move-only type. This will make ownership transfer clear and remove all potential bugs.

This way means PageBuilder always release its page and use OwnedSlice to maintain returned page?
If so,faststring's optimization is useless,because whether array grow happens,it always return a array using new to allocate

gaodayue commented:
> This is an option. But if we use adopt it, user can call finish only one time. I'm not sure if it is convenient for all use scenarios.

I can't think of any use case for the client to call `finish` more than once before `reset`.  The proposed method can be implemented very efficiently almost without memory copy

```
OwnedSlice finish() {
    Slice slice(_buffer);
    slice.data = (char*) _buffer.release();
    return OwnedSlice(slice);
}
```

> This way means PageBuilder always release its page and use OwnedSlice to maintain returned page?

Yes.

> If so,faststring's optimization is useless,because whether array grow happens,it always return a array using new to allocate

`faststring` is not only used by `PageBuilder`, so in other places the small buffer optimization can be useful.

imay commented:
If there is no other use case, I agree with @gaodayue 's proposal. And we should add NOTE to `finish` interface to notice people not to call this function twice.

gaodayue commented:
> My suggestion is to combine Slice finish() and Slice release() into OwnedSlice finish(). OwnedSlice owns slice's data and is a move-only type. This will make ownership transfer clear and remove all potential bugs.

The point is really about use OwnedSlice instead of Slice for ownership transfer to remove potential memory bugs. Whether to keep two versions of `finish` is a less important choice, just follow the KISS principle.

wangbo committed:
(#2037)
1 fix mem fix mem leak when calling string builder.get_dictionary_page;
2 fix delete invalid mem addr in bitshuffleBuilder when no array grow happends
when bitshuffleBuilder didn't grow array, the data page which not use new to allocate will be
returned to ColumnWriter.
When ColumnWriter destructs, the data page will be deleted,this causes core dump

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,60 @@
+class OwnedSlice {
+
+public:
+    Slice slice = nullptr;
```
Slice is not a pointer, better not be assigned from nullptr.
And better to make slice private

wangbo replied:
yep,I meet a core dump,this is a fatal error

wangbo replied:
>And better to make slice private 

to prevent others to change the slice?

imay replied:
Yes, you have already prevented copy and assign, it is reasonable to prevent changing it

wangbo replied:
maybe set OwnedSlice's slice to const is better?

wangbo replied:
when doing move,we just change slice.data, while not slice

imay replied:
slice data is changed, slice is changed too

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,60 @@
+public:
+    Slice slice = nullptr;
+
+    OwnedSlice(){}
```
```suggestion
    OwnedSlice() {}
```

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,60 @@
+
+    OwnedSlice(){}
+
+    OwnedSlice(Slice _slice): slice(_slice){}
```
```suggestion
    OwnedSlice(Slice _slice) : slice(_slice) {}
```

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,60 @@
+    OwnedSlice(Slice _slice): slice(_slice){}
+
+    ~OwnedSlice(){
+        if (slice != nullptr) {
```
Don't compare slice with nullptr

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    OwnedSlice& operator =(OwnedSlice&& src) {
+        if (this != &src) {
```
why not swap this and src?

wangbo replied:
There is assignment operation in std::swap,so call swap in OwnedSlice's move assignment operator would trigger OwnedSlice's move assignment operator again,and then fall into infinite loop

wangbo commented:
major change:
1. remove PageBuilder's finish method, PageBuilder release page to caller;
2. use OwnedSlice to maintain slice produced by PageBuilder

gaodayue commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,68 @@
+class OwnedSlice {
+
+public:
+    Slice slice;
```
Make it private and add a public method to return `const Slice&`.

gaodayue replied:
Make it private and add a public method to return `const Slice&`.

gaodayue commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,68 @@
+    OwnedSlice(Slice _slice) : slice(_slice) {}
+
+    ~OwnedSlice(){
+        if (slice.data != nullptr) {
```
this check is redundant because `delete[] nullptr` is valid

gaodayue commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace doris {
+
+class OwnedSlice {
```
I think you can just declare `OwnedSlice` in slice.h because it's highly related to Slice. In addition, you should add the motivation and usage example as the class comments.

gaodayue commented on be/src/olap/rowset/segment_v2/page_builder.h:
```diff
@@ -75,9 +70,8 @@ class PageBuilder {
-    // and should be followed by reset() before reuse the builder
-    virtual void release() = 0;
+    // This api should be followed by reset() before reuse the builder
+    virtual OwnedSlice release() = 0;
```
`finish()` is a better name and you should move the method declaration right after `add()` and before `reset()`to be consistent with the order of invocation.

gaodayue commented on be/src/olap/rowset/segment_v2/plain_page.h:
```diff
@@ -78,12 +73,14 @@ class PlainPageBuilder : public PageBuilder {
+        encode_fixed32_le((uint8_t *) &_buffer[0], _count);
+        OwnedSlice ownedSlice = OwnedSlice(Slice(_buffer.release(),  PLAIN_PAGE_HEADER_SIZE + _count * SIZE_OF_TYPE));
+
         _buffer.reserve(_options.data_page_size + 1024);
```
move into `reset()`

gaodayue commented on be/src/olap/rowset/segment_v2/plain_page.h:
```diff
@@ -78,12 +73,14 @@ class PlainPageBuilder : public PageBuilder {
         _buffer.reserve(_options.data_page_size + 1024);
-        (void) ret;
+
+        return ownedSlice;
```
```suggestion
        return OwnedSlice(_buffer.release(),  PLAIN_PAGE_HEADER_SIZE + _count * SIZE_OF_TYPE);
```
Add a helper constructor `OwnedSlice(const uint8_t* s, size_t n)` to achieve this.

gaodayue commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,68 @@
+
+    OwnedSlice() {slice.data = nullptr;}
+
+    OwnedSlice(Slice _slice) : slice(_slice) {}
```
```suggestion
    explicit OwnedSlice(Slice slice) : _slice(slice) {}
```
otherwise the following code compiles but will panic at runtime
```
{
  faststring buf(1024);
  OwnedSlice owned_slice(buf);
  // panic: double delete[]
}
```

gaodayue commented on be/src/olap/rowset/segment_v2/binary_dict_page.h:
```diff
@@ -32,6 +32,7 @@
 #include "runtime/mem_tracker.h"
 #include "gen_cpp/segment_v2.pb.h"
 #include "gutil/hash/string_hash.h"
+#include <util/owned_slice.h>
```
Remove this line after `OwnedSlice` is moved into `slice.h`. In addition I want to point out that we use `#include <>` only for standard and third-party library.

gaodayue commented on be/src/olap/rowset/segment_v2/rle_page.h:
```diff
@@ -122,11 +113,14 @@ class RlePageBuilder : public PageBuilder {
+        // or it will lead to a bug if the header is less than 8 byte and the data is small
+        _rle_encoder->Flush();
+        encode_fixed32_le(&_buf[0], _count);
+        return OwnedSlice(Slice(_buf.release(), _buf.size()));
```
`_buf.size()` is 0 after `_buf.release()`. You should save `_buf.size()` first.

wangbo committed:
update commit

wangbo committed:
move buffer.reserve from finish to reset in plain_page

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,82 @@
+        if (this != &src) {
+            std::swap(_slice, src._slice);
+            delete[] src._slice.data;
+            src._slice.size = 0;
```
no need to delete[] and set size to zero here. Just leave it to the input OwnedSlice to delete it.

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,82 @@
+class OwnedSlice {
+
+public:
+    OwnedSlice() {_slice.data = nullptr;_slice.size=0;}
```
```suggestion
    OwnedSlice() : _slice((uint8_t*)nullptr, 0) {}
```

wangbo replied:
👌

imay commented on be/src/util/owned_slice.h:
```diff
@@ -0,0 +1,82 @@
+        return _slice;
+    }
+
+    private:
```
```suggestion
private:
```

wangbo replied:
👌

imay commented on be/src/olap/rowset/segment_v2/column_writer.h:
```diff
@@ -102,14 +102,13 @@ class ColumnWriter {
-        Slice null_bitmap = Slice((uint8_t*)nullptr, 0);
-        Slice data = Slice((uint8_t*)nullptr, 0);
+        Slice null_bitmap = Slice((const uint8_t*)nullptr, 0);
+        OwnedSlice data = OwnedSlice((uint8_t*)nullptr, 0);
```
why not keep null_bitmap and data use same type?

wangbo replied:
null_bitmap produced by NullBitmapBuilder,and NullBitmapBuilder didn't implement PageBuilder,here just keep NullBitmapBuilder's origin Semantics.

wangbo committed:
update owned_slice.h

imay reviewed (approved):
LGTM
